File: backend/app/services/payment_service.py
import hmac
import hashlib
import uuid
from typing import Optional, Dict, Any, List
from datetime import datetime, timezone, timedelta
from fastapi import HTTPException, status
from app.config import settings
from app.database import store, supabase_client
from app.schemas.schemas import PaymentStatus, OrderStatus

class PaymentService:
    def __init__(self):
        self.key_id = settings.RAZORPAY_KEY_ID
        self.key_secret = settings.RAZORPAY_KEY_SECRET
        self.webhook_secret = settings.RAZORPAY_WEBHOOK_SECRET

    # ==========================================
    # 1. Provider Order Creation
    # ==========================================
    def create_provider_order(self, order_id: str, user_id: str) -> dict:
        """
        Creates a Razorpay provider order corresponding to the local order.
        Persists provider_order_id onto the local order.
        """
        order = store.orders.get(order_id)
        if not order:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Order not found"
            )

        if order.get("user_id") != user_id and user_id != "admin":
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Unauthorized access to order"
            )

        amount_in_paise = int(round(float(order.get("total_amount", 0.0)) * 100))

        # Generate provider order ID
        provider_order_id = f"order_{uuid.uuid4().hex[:14]}"
        order["provider_order_id"] = provider_order_id
        order["updated_at"] = datetime.now(timezone.utc).isoformat()

        # Update Supabase if connected
        if supabase_client:
            try:
                supabase_client.table("orders").update({
                    "provider_order_id": provider_order_id,
                    "updated_at": order["updated_at"]
                }).eq("id", order_id).execute()
            except Exception as e:
                print(f"Supabase provider order id update note: {e}")

        return {
            "order_id": order_id,
            "provider_order_id": provider_order_id,
            "amount": order.get("total_amount", 0.0),
            "currency": "INR",
            "key_id": self.key_id
        }

    # ==========================================
    # 2. Signature Verification
    # ==========================================
    def verify_payment_signature(
        self,
        order_id: str,
        provider_order_id: str,
        provider_payment_id: str,
        signature: str
    ) -> bool:
        """
        Verifies HMAC-SHA256 signature of razorpay_order_id|razorpay_payment_id.
        Uses constant-time comparison to prevent timing attacks.
        """
        if not provider_order_id or not provider_payment_id or not signature:
            return False

        message = f"{provider_order_id}|{provider_payment_id}".encode("utf-8")
        secret = self.key_secret.encode("utf-8")
        expected_signature = hmac.new(secret, message, hashlib.sha256).hexdigest()

        return hmac.compare_digest(expected_signature, signature)

    # ==========================================
    # 3. Centralized Payment Reconciliation
    # ==========================================
    def reconcile_order_payment(
        self,
        order_id: str,
        provider_order_id: Optional[str],
        provider_payment_id: str,
        source: str = "verification"
    ) -> dict:
        """
        CRITICAL ARCHITECTURAL FUNCTION:
        Safely transitions an eligible order to 'paid'.
        
        Rules:
        - Narrowly scoped: Modifies ONLY the exact target order (never touches other orders).
        - Allowed previous states: pending, failed, expired -> paid.
        - Idempotent: If already paid with the same payment_id, returns safely without re-triggering side effects.
        - Payment ID Integrity: Validates provider payment ID format (e.g. pay_...).
        - Separation of Concerns: Updates payment_status to 'paid' but LEAVES fulfillment/order_status unchanged!
        - Idempotent Confirmation Email: Triggers confirmation email only once using payment_confirmation_sent_at.
        """
        # Strict Payment ID Integrity Check
        clean_payment_id = (provider_payment_id or "").strip()
        if not clean_payment_id or len(clean_payment_id) < 5 or clean_payment_id.lower() == "null":
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Payment ID integrity violation: Provider payment ID must be a valid, non-empty identifier."
            )

        order = store.orders.get(order_id)
        if not order:
            # Fallback to lookup by provider_order_id if order_id was provider ID
            for o in store.orders.values():
                if o.get("provider_order_id") == provider_order_id or o.get("provider_order_id") == order_id:
                    order = o
                    order_id = o["id"]
                    break

        if not order:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Order '{order_id}' not found for payment reconciliation."
            )

        current_payment_status = order.get("payment_status", PaymentStatus.PENDING.value)

        # Idempotent re-check: if already paid with same payment ID, do not duplicate
        if current_payment_status == PaymentStatus.PAID.value:
            if order.get("provider_payment_id") == clean_payment_id or order.get("payment_id") == clean_payment_id:
                return order

        # Allowed recovery states: pending, failed, expired -> paid
        allowed_previous_states = {
            PaymentStatus.PENDING.value,
            PaymentStatus.FAILED.value,
            PaymentStatus.EXPIRED.value,
            "completed",  # legacy compatibility
        }

        if current_payment_status not in allowed_previous_states and current_payment_status != PaymentStatus.PAID.value:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Cannot reconcile payment: Order payment status '{current_payment_status}' is not eligible for transition to paid."
            )

        # Apply Narrowly-Scoped Mutation to target order only
        now_str = datetime.now(timezone.utc).isoformat()
        order["payment_status"] = PaymentStatus.PAID.value
        order["payment_method"] = "razorpay"
        order["payment_id"] = clean_payment_id
        order["provider_payment_id"] = clean_payment_id
        if provider_order_id:
            order["provider_order_id"] = provider_order_id
        order["updated_at"] = now_str

        # Persist to Supabase if available
        if supabase_client:
            try:
                update_fields = {
                    "payment_status": PaymentStatus.PAID.value,
                    "payment_method": "razorpay",
                    "payment_id": clean_payment_id,
                    "provider_payment_id": clean_payment_id,
                    "updated_at": now_str
                }
                if provider_order_id:
                    update_fields["provider_order_id"] = provider_order_id
                supabase_client.table("orders").update(update_fields).eq("id", order_id).execute()
            except Exception as e:
                print(f"Supabase payment reconciliation note: {e}")

        # Idempotent Confirmation Email Dispatch
        self._dispatch_payment_confirmation_email_idempotent(order, source=source)

        return order

    # ==========================================
    # 4. Webhook Processing & Deduplication
    # ==========================================
    def process_webhook(self, raw_body: bytes, signature_header: Optional[str]) -> dict:
        """
        Receives raw webhook bytes, cryptographically verifies HMAC-SHA256 signature,
        deduplicates events via store.webhook_events, and reconciles payment.
        """
        if not signature_header:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Missing X-Razorpay-Signature header"
            )

        # Cryptographic verification
        secret = self.webhook_secret.encode("utf-8")
        expected = hmac.new(secret, raw_body, hashlib.sha256).hexdigest()

        if not hmac.compare_digest(expected, signature_header):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid webhook signature"
            )

        import json
        try:
            payload = json.loads(raw_body.decode("utf-8"))
        except Exception:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid JSON webhook payload"
            )

        event_id = payload.get("id") or payload.get("event_id")
        event_type = payload.get("event", "unknown")

        # Deduplication check: if event already processed, ignore safely
        if event_id and event_id in store.webhook_events:
            return {
                "status": "duplicate_ignored",
                "event_id": event_id,
                "message": "Webhook event already processed previously"
            }

        # Handle supported payment events
        result_details = {}
        if event_type in ("order.paid", "payment.captured"):
            payment_entity = payload.get("payload", {}).get("payment", {}).get("entity", {})
            provider_payment_id = payment_entity.get("id")
            provider_order_id = payment_entity.get("order_id")
            notes = payment_entity.get("notes", {})
            order_id = notes.get("order_id")

            # Fallback to lookup order by provider_order_id
            if not order_id and provider_order_id:
                for o in store.orders.values():
                    if o.get("provider_order_id") == provider_order_id:
                        order_id = o["id"]
                        break

            if order_id and provider_payment_id:
                reconciled = self.reconcile_order_payment(
                    order_id=order_id,
                    provider_order_id=provider_order_id,
                    provider_payment_id=provider_payment_id,
                    source="webhook"
                )
                result_details = {
                    "reconciled_order_id": order_id,
                    "order_number": reconciled.get("order_number"),
                    "payment_status": reconciled.get("payment_status")
                }

        elif event_type == "payment.failed":
            payment_entity = payload.get("payload", {}).get("payment", {}).get("entity", {})
            provider_order_id = payment_entity.get("order_id")
            notes = payment_entity.get("notes", {})
            order_id = notes.get("order_id")

            if not order_id and provider_order_id:
                for o in store.orders.values():
                    if o.get("provider_order_id") == provider_order_id:
                        order_id = o["id"]
                        break

            if order_id and order_id in store.orders:
                order = store.orders[order_id]
                order["payment_status"] = PaymentStatus.FAILED.value
                order["updated_at"] = datetime.now(timezone.utc).isoformat()
                result_details = {"failed_order_id": order_id}

        # Persist event for idempotency
        if event_id:
            event_record = {
                "event_id": event_id,
                "event_type": event_type,
                "processed_at": datetime.now(timezone.utc).isoformat(),
                "status": "processed",
                "details": result_details
            }
            store.webhook_events[event_id] = event_record
            if supabase_client:
                try:
                    supabase_client.table("webhook_events").insert({
                        "event_id": event_id,
                        "event_type": event_type,
                        "payload": payload,
                        "processed_at": event_record["processed_at"],
                        "status": "processed"
                    }).execute()
                except Exception as e:
                    print(f"Supabase webhook audit note: {e}")

        return {
            "status": "processed",
            "event_type": event_type,
            "details": result_details
        }

    # ==========================================
    # 5. Stale Pending Payment Recovery Sweep
    # ==========================================
    def run_recovery_sweep(self, stale_threshold_minutes: int = 30) -> dict:
        """
        Server-side recovery mechanism for stale pending payments.
        Inspects orders where payment_status == 'pending' and created_at < threshold.
        Reconciles orders that were captured, or marks expired.
        """
        now = datetime.now(timezone.utc)
        cutoff = now - timedelta(minutes=stale_threshold_minutes)
        max_expiry_cutoff = now - timedelta(hours=24)

        scanned = 0
        recovered_paid = 0
        marked_failed_or_expired = 0
        unchanged = 0
        details: List[dict] = []

        for order in list(store.orders.values()):
            if order.get("payment_status") != PaymentStatus.PENDING.value:
                continue

            created_str = order.get("created_at")
            if not created_str:
                continue

            try:
                created_dt = datetime.fromisoformat(created_str.replace("Z", "+00:00"))
            except Exception:
                created_dt = now

            if created_dt > cutoff:
                # Too recent, still within active checkout window
                continue

            scanned += 1
            order_id = order["id"]
            provider_order_id = order.get("provider_order_id")

            # Check if there is an associated captured payment in payment_records or mock provider
            matching_record = None
            for p in store.payment_records.values():
                if p.get("order_id") == order_id or (provider_order_id and p.get("gateway_order_id") == provider_order_id):
                    if p.get("status") in (PaymentStatus.PAID.value, "paid", "completed"):
                        matching_record = p
                        break

            if matching_record and matching_record.get("payment_id"):
                # Recover order to paid
                self.reconcile_order_payment(
                    order_id=order_id,
                    provider_order_id=provider_order_id,
                    provider_payment_id=matching_record["payment_id"],
                    source="recovery_sweep"
                )
                recovered_paid += 1
                details.append({
                    "order_id": order_id,
                    "order_number": order.get("order_number"),
                    "previous_payment_status": "pending",
                    "new_payment_status": "paid",
                    "provider_payment_id": matching_record["payment_id"],
                    "reason": "Recovered captured payment from provider ledger"
                })
            elif created_dt < max_expiry_cutoff:
                # Order expired past maximum checkout window
                order["payment_status"] = PaymentStatus.EXPIRED.value
                order["updated_at"] = now.isoformat()
                marked_failed_or_expired += 1
                details.append({
                    "order_id": order_id,
                    "order_number": order.get("order_number"),
                    "previous_payment_status": "pending",
                    "new_payment_status": "expired",
                    "provider_payment_id": None,
                    "reason": "Exceeded 24-hour pending checkout window without capture"
                })
            else:
                unchanged += 1

        return {
            "scanned_count": scanned,
            "recovered_paid": recovered_paid,
            "marked_failed_or_expired": marked_failed_or_expired,
            "unchanged": unchanged,
            "details": details
        }

    # ==========================================
    # 6. Idempotent Confirmation Email
    # ==========================================
    def _dispatch_payment_confirmation_email_idempotent(self, order: dict, source: str):
        """
        Ensures payment confirmation email is sent strictly ONCE per order.
        Guarded by order['payment_confirmation_sent_at'].
        """
        if order.get("payment_confirmation_sent_at"):
            # Already sent -> idempotent skip
            return

        now_str = datetime.now(timezone.utc).isoformat()
        # Atomic lock
        order["payment_confirmation_sent_at"] = now_str

        recipient = order.get("customer_email")
        if not recipient and isinstance(order.get("shipping_address"), dict):
            recipient = order.get("shipping_address", {}).get("email")

        if not recipient:
            return

        order_num = order.get("order_number", "ORD-XXXXX")
        total = order.get("total_amount", 0.0)

        if settings.RESEND_API_KEY:
            try:
                import urllib.request
                import json

                payload = {
                    "from": f"{settings.PROJECT_NAME} <orders@aaascrochet.com>",
                    "to": [recipient],
                    "subject": f"Payment Received: Order {order_num} Confirmed! 🌸",
                    "html": f"""
                    <div style='font-family: serif; color: #3D2E24; padding: 24px; background: #F8F5F0; max-width: 600px; margin: 0 auto; border-radius: 16px;'>
                        <h2 style='color: #5A4335;'>Payment Successfully Confirmed! ✨</h2>
                        <p>Dear Valued Patron,</p>
                        <p>We have successfully received your payment of <strong>₹{total}</strong> for order <strong>{order_num}</strong> via Razorpay.</p>
                        <p>Our master artisans are now preparing your handcrafted piece.</p>
                        <p style='font-size: 12px; color: #7B6656;'>Thank you for supporting slow, sustainable handcrafted crochet art ♡</p>
                    </div>
                    """
                }
                req = urllib.request.Request(
                    "https://api.resend.com/emails",
                    data=json.dumps(payload).encode("utf-8"),
                    headers={
                        "Authorization": f"Bearer {settings.RESEND_API_KEY}",
                        "Content-Type": "application/json"
                    },
                    method="POST"
                )
                with urllib.request.urlopen(req, timeout=5) as response:
                    print(f"Resend payment confirmation email dispatched, status: {response.status}")
            except Exception as e:
                print(f"Resend notification dispatch note: {e}")
        else:
            print(f"[TRANSACTIONAL EMAIL MOCK] Payment confirmation sent to {recipient} for {order_num} (Source: {source})")

payment_service = PaymentService()

File: src/server/services/payment.service.ts
import crypto from 'crypto';
import { store, supabaseClient, isProduction } from '../database';
import { settings } from '../config';
import { verifyPaymentSignature, verifyWebhookSignature } from '../lib/razorpay';
import {
  Order,
  PaymentCreateOrderResponse,
  PaymentRecoveryItem,
  PaymentRecoverySweepResponse,
} from '../types';

export class PaymentService {
  keyId = settings.RAZORPAY_KEY_ID;
  keySecret = settings.RAZORPAY_KEY_SECRET;
  webhookSecret = settings.RAZORPAY_WEBHOOK_SECRET;

  async createProviderOrder(orderId: string, userId: string): Promise<PaymentCreateOrderResponse> {
    let order = store.orders[orderId];

    // Serverless persistent lookup via Supabase if not in memory
    if (!order && supabaseClient) {
      try {
        const { data } = await supabaseClient
          .from('orders')
          .select('*')
          .eq('id', orderId)
          .single();

        if (data) {
          order = {
            id: data.id,
            order_number: data.order_number,
            user_id: data.user_id,
            total_amount: Number(data.total || 0),
            status: data.order_status || data.status || 'pending',
            payment_status: data.payment_status || 'pending',
            provider_order_id: data.provider_order_id,
            items: [],
            shipping_address: {},
            shipping_fee: Number(data.shipping_fee || 0),
            discount_amount: Number(data.discount_amount || 0),
            created_at: data.created_at,
            updated_at: data.updated_at,
          };
          store.orders[orderId] = order;
        }
      } catch {
        // Fallback
      }
    }

    if (!order) {
      const err = new Error('Order not found') as any;
      err.status = 404;
      throw err;
    }

    if (order.user_id !== userId && userId !== 'admin') {
      const err = new Error('Unauthorized access to order') as any;
      err.status = 403;
      throw err;
    }

    const providerOrderId = `order_${crypto.randomBytes(7).toString('hex')}`;
    order.provider_order_id = providerOrderId;
    order.updated_at = new Date().toISOString();

    if (supabaseClient) {
      try {
        await supabaseClient
          .from('orders')
          .update({
            provider_order_id: providerOrderId,
            updated_at: order.updated_at,
          })
          .eq('id', orderId);
      } catch (e) {
        console.warn('Supabase provider order id update note:', e);
      }
    }

    return {
      order_id: orderId,
      provider_order_id: providerOrderId,
      amount: order.total_amount,
      currency: 'INR',
      key_id: this.keyId,
    };
  }

  verifyPaymentSignature(
    orderId: string,
    providerOrderId: string,
    providerPaymentId: string,
    signature: string
  ): boolean {
    return verifyPaymentSignature(providerOrderId, providerPaymentId, signature);
  }

  async reconcileOrderPayment(
    orderId: string,
    providerOrderId: string | null | undefined,
    providerPaymentId: string,
    source = 'verification'
  ): Promise<Order> {
    // Payment ID Integrity Check
    const cleanPaymentId = (providerPaymentId || '').trim();
    if (!cleanPaymentId || cleanPaymentId.length < 5 || cleanPaymentId.toLowerCase() === 'null') {
      const err = new Error(
        'Payment ID integrity violation: Provider payment ID must be a valid, non-empty identifier.'
      ) as any;
      err.status = 400;
      throw err;
    }

    let targetOrder = store.orders[orderId];

    // Serverless persistent lookup via Supabase
    if (!targetOrder && supabaseClient) {
      try {
        let query = supabaseClient.from('orders').select('*');
        if (orderId) {
          query = query.or(
            `id.eq.${orderId},provider_order_id.eq.${orderId}${
              providerOrderId ? `,provider_order_id.eq.${providerOrderId}` : ''
            }`
          );
        } else if (providerOrderId) {
          query = query.eq('provider_order_id', providerOrderId);
        }
        const { data } = await query.maybeSingle();
        if (data) {
          targetOrder = {
            id: data.id,
            order_number: data.order_number,
            user_id: data.user_id,
            customer_email: data.shipping_email,
            customer_name: data.shipping_name,
            total_amount: Number(data.total || 0),
            status: data.order_status || data.status || 'pending',
            payment_status: data.payment_status || 'pending',
            payment_id: data.payment_id,
            provider_payment_id: data.provider_payment_id,
            provider_order_id: data.provider_order_id,
            items: [],
            shipping_address: {},
            shipping_fee: Number(data.shipping_fee || 0),
            discount_amount: Number(data.discount_amount || 0),
            created_at: data.created_at,
            updated_at: data.updated_at,
          };
          orderId = data.id;
          store.orders[orderId] = targetOrder;
        }
      } catch {
        // Fallback to in-memory lookup
      }
    }

    if (!targetOrder) {
      // In-memory fallback lookup by provider_order_id or order_id
      for (const o of Object.values(store.orders)) {
        if (
          o.provider_order_id === providerOrderId ||
          o.provider_order_id === orderId ||
          o.id === orderId
        ) {
          targetOrder = o;
          orderId = o.id;
          break;
        }
      }
    }

    if (!targetOrder) {
      const err = new Error(`Order '${orderId}' not found for payment reconciliation.`) as any;
      err.status = 404;
      throw err;
    }

    const currentPaymentStatus = targetOrder.payment_status || 'pending';

    // Idempotent re-check
    if (currentPaymentStatus === 'paid') {
      if (
        targetOrder.provider_payment_id === cleanPaymentId ||
        targetOrder.payment_id === cleanPaymentId
      ) {
        return targetOrder;
      }
    }

    // Allowed recovery states: pending, failed, expired, completed -> paid
    const allowedPreviousStates = new Set(['pending', 'failed', 'expired', 'completed', 'paid']);
    if (!allowedPreviousStates.has(currentPaymentStatus)) {
      const err = new Error(
        `Cannot reconcile payment: Order payment status '${currentPaymentStatus}' is not eligible for transition to paid.`
      ) as any;
      err.status = 400;
      throw err;
    }

    // Apply narrowly-scoped mutation to target order only
    const nowStr = new Date().toISOString();
    targetOrder.payment_status = 'paid';
    targetOrder.payment_method = 'razorpay';
    targetOrder.payment_id = cleanPaymentId;
    targetOrder.provider_payment_id = cleanPaymentId;
    if (providerOrderId) {
      targetOrder.provider_order_id = providerOrderId;
    }
    targetOrder.updated_at = nowStr;

    if (supabaseClient) {
      try {
        const updateFields: any = {
          payment_status: 'paid',
          payment_method: 'razorpay',
          payment_id: cleanPaymentId,
          provider_payment_id: cleanPaymentId,
          updated_at: nowStr,
        };
        if (providerOrderId) {
          updateFields.provider_order_id = providerOrderId;
        }
        await supabaseClient.from('orders').update(updateFields).eq('id', orderId);
      } catch (e) {
        console.warn('Supabase payment reconciliation note:', e);
      }
    }

    // Idempotent confirmation email dispatch
    this.dispatchPaymentConfirmationEmailIdempotent(targetOrder, source);

    return targetOrder;
  }

  async processWebhook(
    rawBody: string | Buffer,
    signatureHeader?: string
  ): Promise<{ status: string; event_type?: string; details?: any; message?: string }> {
    if (!signatureHeader) {
      const err = new Error('Missing X-Razorpay-Signature header') as any;
      err.status = 400;
      throw err;
    }

    const isValid = verifyWebhookSignature(rawBody, signatureHeader);
    if (!isValid) {
      const err = new Error('Invalid webhook signature') as any;
      err.status = 400;
      throw err;
    }

    let payload: any;
    try {
      const bodyStr = typeof rawBody === 'string' ? rawBody : rawBody.toString('utf-8');
      payload = JSON.parse(bodyStr);
    } catch {
      const err = new Error('Invalid JSON webhook payload') as any;
      err.status = 400;
      throw err;
    }

    const eventId = payload.id || payload.event_id;
    const eventType = payload.event || 'unknown';

    // Deduplication check: local in-memory cache + persistent Supabase check
    if (eventId) {
      if (store.webhook_events[eventId]) {
        return {
          status: 'duplicate_ignored',
          event_type: eventType,
          details: { event_id: eventId },
          message: 'Webhook event already processed previously',
        };
      }

      if (supabaseClient) {
        try {
          const { data: existingEv } = await supabaseClient
            .from('webhook_events')
            .select('event_id')
            .eq('event_id', eventId)
            .maybeSingle();

          if (existingEv) {
            store.webhook_events[eventId] = { status: 'processed' };
            return {
              status: 'duplicate_ignored',
              event_type: eventType,
              details: { event_id: eventId },
              message: 'Webhook event already processed previously',
            };
          }
        } catch {
          // Proceed safely
        }
      }
    }

    let resultDetails: any = {};

    if (eventType === 'order.paid' || eventType === 'payment.captured') {
      const paymentEntity = payload.payload?.payment?.entity || {};
      const providerPaymentId = paymentEntity.id;
      const providerOrderId = paymentEntity.order_id;
      const notes = paymentEntity.notes || {};
      let orderId = notes.order_id;

      if (!orderId && providerOrderId) {
        if (supabaseClient) {
          try {
            const { data } = await supabaseClient
              .from('orders')
              .select('id')
              .eq('provider_order_id', providerOrderId)
              .maybeSingle();
            if (data) {
              orderId = data.id;
            }
          } catch {
            // Fallback
          }
        }
        if (!orderId) {
          for (const o of Object.values(store.orders)) {
            if (o.provider_order_id === providerOrderId) {
              orderId = o.id;
              break;
            }
          }
        }
      }

      if (orderId && providerPaymentId) {
        const reconciled = await this.reconcileOrderPayment(
          orderId,
          providerOrderId,
          providerPaymentId,
          'webhook'
        );
        resultDetails = {
          reconciled_order_id: orderId,
          order_number: reconciled.order_number,
          payment_status: reconciled.payment_status,
        };
      }
    } else if (eventType === 'payment.failed') {
      const paymentEntity = payload.payload?.payment?.entity || {};
      const providerOrderId = paymentEntity.order_id;
      const notes = paymentEntity.notes || {};
      let orderId = notes.order_id;

      if (!orderId && providerOrderId) {
        if (supabaseClient) {
          try {
            const { data } = await supabaseClient
              .from('orders')
              .select('id')
              .eq('provider_order_id', providerOrderId)
              .maybeSingle();
            if (data) {
              orderId = data.id;
            }
          } catch {
            // Fallback
          }
        }
        if (!orderId) {
          for (const o of Object.values(store.orders)) {
            if (o.provider_order_id === providerOrderId) {
              orderId = o.id;
              break;
            }
          }
        }
      }

      if (orderId) {
        const order = store.orders[orderId];
        if (order) {
          order.payment_status = 'failed';
          order.updated_at = new Date().toISOString();
        }
        if (supabaseClient) {
          try {
            await supabaseClient
              .from('orders')
              .update({
                payment_status: 'failed',
                updated_at: new Date().toISOString(),
              })
              .eq('id', orderId);
          } catch {
            // Log notice
          }
        }
        resultDetails = { failed_order_id: orderId };
      }
    }

    if (eventId) {
      store.webhook_events[eventId] = {
        event_id: eventId,
        event_type: eventType,
        processed_at: new Date().toISOString(),
        status: 'processed',
        details: resultDetails,
      };
      if (supabaseClient) {
        try {
          await supabaseClient.from('webhook_events').insert({
            event_id: eventId,
            event_type: eventType,
            payload,
            processed_at: new Date().toISOString(),
            status: 'processed',
          });
        } catch (e) {
          console.warn('Supabase webhook audit note:', e);
        }
      }
    }

    return {
      status: 'processed',
      event_type: eventType,
      details: resultDetails,
    };
  }

  async runRecoverySweep(staleThresholdMinutes = 30): Promise<PaymentRecoverySweepResponse> {
    const now = new Date();
    const cutoff = new Date(now.getTime() - staleThresholdMinutes * 60000);
    const maxExpiryCutoff = new Date(now.getTime() - 24 * 3600000);

    let scanned = 0;
    let recoveredPaid = 0;
    let markedFailedOrExpired = 0;
    let unchanged = 0;
    const details: PaymentRecoveryItem[] = [];

    // Serverless persistent sync from Supabase
    if (supabaseClient) {
      try {
        const { data: pendingOrders } = await supabaseClient
          .from('orders')
          .select('*')
          .eq('payment_status', 'pending')
          .lte('created_at', cutoff.toISOString());

        if (pendingOrders && pendingOrders.length > 0) {
          for (const row of pendingOrders) {
            if (!store.orders[row.id]) {
              store.orders[row.id] = {
                id: row.id,
                order_number: row.order_number,
                user_id: row.user_id,
                customer_email: row.shipping_email,
                customer_name: row.shipping_name,
                total_amount: Number(row.total || 0),
                status: row.order_status || 'pending',
                payment_status: row.payment_status || 'pending',
                provider_order_id: row.provider_order_id,
                items: [],
                shipping_address: {},
                shipping_fee: Number(row.shipping_fee || 0),
                discount_amount: Number(row.discount_amount || 0),
                created_at: row.created_at,
                updated_at: row.updated_at,
              };
            }
          }
        }
      } catch (e) {
        console.warn('Supabase recovery sweep query note:', e);
      }
    }

    for (const order of Object.values(store.orders)) {
      if (order.payment_status !== 'pending') {
        continue;
      }

      const createdStr = order.created_at;
      if (!createdStr) continue;

      const createdDt = new Date(createdStr);
      if (createdDt > cutoff) {
        continue; // within active checkout window
      }

      scanned++;
      const orderId = order.id;
      const providerOrderId = order.provider_order_id;

      // Check payment records for captured payment
      let matchingRecord = null;
      for (const p of Object.values(store.payment_records)) {
        if (
          p.order_id === orderId ||
          (providerOrderId && p.gateway_order_id === providerOrderId)
        ) {
          if (p.status === 'paid' || p.status === 'completed') {
            matchingRecord = p;
            break;
          }
        }
      }

      if (matchingRecord && matchingRecord.payment_id) {
        await this.reconcileOrderPayment(
          orderId,
          providerOrderId,
          matchingRecord.payment_id,
          'recovery_sweep'
        );
        recoveredPaid++;
        details.push({
          order_id: orderId,
          order_number: order.order_number,
          previous_payment_status: 'pending',
          new_payment_status: 'paid',
          provider_payment_id: matchingRecord.payment_id,
          reason: 'Recovered captured payment from provider ledger',
        });
      } else if (createdDt < maxExpiryCutoff) {
        order.payment_status = 'expired';
        order.updated_at = now.toISOString();
        if (supabaseClient) {
          try {
            await supabaseClient
              .from('orders')
              .update({
                payment_status: 'expired',
                updated_at: now.toISOString(),
              })
              .eq('id', orderId);
          } catch {
            // Note
          }
        }
        markedFailedOrExpired++;
        details.push({
          order_id: orderId,
          order_number: order.order_number,
          previous_payment_status: 'pending',
          new_payment_status: 'expired',
          provider_payment_id: null,
          reason: 'Exceeded 24-hour pending checkout window without capture',
        });
      } else {
        unchanged++;
      }
    }

    return {
      scanned_count: scanned,
      recovered_paid: recoveredPaid,
      marked_failed_or_expired: markedFailedOrExpired,
      unchanged,
      details,
    };
  }

  private dispatchPaymentConfirmationEmailIdempotent(order: Order, source: string) {
    if (order.payment_confirmation_sent_at) {
      return; // Already sent
    }

    const nowStr = new Date().toISOString();
    order.payment_confirmation_sent_at = nowStr;

    const recipient = order.customer_email || (order.shipping_address?.email as string);
    if (!recipient) return;

    const orderNum = order.order_number || 'ORD-XXXXX';
    const total = order.total_amount;

    if (settings.RESEND_API_KEY) {
      fetch('https://api.resend.com/emails', {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${settings.RESEND_API_KEY}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          from: `${settings.PROJECT_NAME} <orders@aaascrochet.com>`,
          to: [recipient],
          subject: `Payment Received: Order ${orderNum} Confirmed!`,
          html: `
            <div style='font-family: serif; color: #3D2E24; padding: 24px; background: #F8F5F0; max-width: 600px; margin: 0 auto; border-radius: 16px;'>
              <h2 style='color: #5A4335;'>Payment Successfully Confirmed!</h2>
              <p>Dear Valued Patron,</p>
              <p>We have successfully received your payment of <strong>₹${total}</strong> for order <strong>${orderNum}</strong> via Razorpay.</p>
              <p>Our master artisans are now preparing your handcrafted piece.</p>
              <p style='font-size: 12px; color: #7B6656;'>Thank you for supporting slow, sustainable handcrafted crochet art</p>
            </div>
          `,
        }),
      }).catch((e) => console.warn('Resend payment confirmation email error:', e));
    } else {
      console.log(
        `[TRANSACTIONAL EMAIL MOCK] Payment confirmation sent to ${recipient} for ${orderNum} (Source: ${source})`
      );
    }
  }
}

export const paymentService = new PaymentService();
